[AprupKale] iP

AprupKale opened the pull request:
# Max
## your most efficient manager 🏅 

> “Your mind is for having ideas, not holding them.” – David Allen

Max is
- easy to learn
- super fast
- SUPER *AWESOME*

What you need to do is
- [ ] fork this repo from [here](https://github.com/AprupKale/ip/)
- [ ] clone your fork locally
- [ ] run the bot locally
- [x] HAVE A FUN TIME BEING *ORGANIZED*

What kind of tasks can you add
1. Todo
2. Deadline
3. Event 

**THE BEST PART IS IT IS FREE**

All you need to do to run the bot is call `Duke.main();`

```java
public static void main(String[] args) {
    Duke max = new Duke("data/duke.txt");
    max.run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

AprupKale committed:
Ignore .class files

AprupKale committed:
Update for Level 0 - Add greet and exit

AprupKale committed:
Update for Level 1 - Add user input echo

AprupKale committed:
Update for Level 2 - Add list to store user input

AprupKale committed:
Update for Level 3 - Add Task and ability to mark and unmark tasks done

AprupKale committed:
Remove redundant code in greet and exit

AprupKale committed:
Update for Level 4 - Add Todo, Deadline and Event

AprupKale committed:
Update for A-TextUiTesting

AprupKale committed:
Update for Level 5 - Add exception handling with DukeException

AprupKale committed:
Update for Level 6 - Add ability to delete tasks

AprupKale committed:
Change List to TaskList

AprupKale committed:
Update for Level 7 - Add save ability

AprupKale committed:
Merge branch 'branch-Level-7'

AprupKale committed:
Modify text-ui-test/input.txt and text-ui-test/EXPECTED.TXT

AprupKale committed:
Update for Level 8 - Add date and time format for deadlines and events

AprupKale committed:
Merge branch 'branch-Level-8'

AprupKale committed:
Update for A-MoreOOP - Add Parser, Storage and Ui classes

AprupKale committed:
Merge branch 'branch-A-MoreOOP'

AprupKale committed:
Update for A-Packages - Add packages duke, duke.tasks, and duke.exceptions

AprupKale committed:
Add DukeIndexOutOfBoundsException, DukeParseException, and DukeDateTimeParseException

AprupKale committed:
Merge branch 'branch-A-Packages'

AprupKale committed:
Merge branch 'master' into add-gradle-support

AprupKale committed:
Merge branch 'add-gradle-support'

AprupKale committed:
Update for A-Gradle - Add Gradle to build and run tests on project

AprupKale committed:
Add equals() method to Task, Todo, Deadline, Event and TaskList

AprupKale committed:
Update for A-JUnit - Add Test classes for Parser and TaskList

AprupKale committed:
Update for A-Jar - Edit build.gradle to build .jar file

AprupKale committed:
Update for A-JavaDoc - Add documentation for all public fields and methods

AprupKale committed:
Update for A-CodingStandard - Add config/checkstyle/checkstyle.xml and config/checkstyle/suppressions.xml

AprupKale committed:
Update for Level 9 - Add functionality to search task with a keyword

AprupKale committed:
Fix coding standard errors

AprupKale committed:
Merge branch 'branch-A-CodingStandard'

AprupKale committed:
Fix conflicts in Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Storage.java

AprupKale committed:
Merge branch 'branch-Level-9'

AprupKale committed:
Add documentation for Task::containsKeyword and Parser::getTaskList

songfangyl commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,102 @@
+            } else {
+                return false;
+            }
+        }
```
Is there any other way to do this to avoid this deep nesting?
Here is my suggestion:

```
if (taskList.getNumberOfTasks() == this.getNumberOfTasks()) {
    return false;
}
for (int i = 0...) {
    ....
}
```

songfangyl commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,115 @@
+     * @param storage Storage object to interact with data file.
+     * @throws DukeException If any exception is encountered.
+     */
+    public void getUserInput(TaskList taskList, Storage storage) throws DukeException {
```
I think the name of this method is not so accurate, because when I see "get" in method name, I always expect this method return something to me, but this method does not return anything 🤔

songfangyl commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,115 @@
+
+            storage.writeToFile(taskList);
+        }
+    }
```
It is better to split this method into multiple methods, because this method not only get user input, but also display the message. My suggestion is the display message can be done in another method, an example shows below

```ruby
public void printAddTaskMessage(Task task, TaskList taskList) {
        System.out.println("Got it, I've added this task:\n    "  +
                task.toString() + "\n" +
                "Now you have " + taskList.size() + " tasks in the list.");
}
```

songfangyl commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,115 @@
+            System.out.print("-");
+        }
+        System.out.println("");
+    }
```
I like the way you print the line! I just realized I am hard coding the line🤣

songfangyl commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    /**
+     * Return status of task as Integer.
```
Should be "Returns" instead of "Return" right?

songfangyl commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,115 @@
+    /**
+     * Displays exit message.
+     */
+    public void exit() {
```
Is it better to have a method name like showExitMessage? 🤔

songfangyl reviewed (commented):
Overall is good, I think for the UI object, it did more than just interact with the user, which is not so accurate, but I am inspired by your code, after reveiew your code, I realized that I got a lot of things can be improved! 😃

yezkez10 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,101 @@
+                    break;
+
+                case "D":
+                    list.addToList(new Deadline(params[2], isDone, LocalDateTime.parse(params[3], Duke.TIME_FORMAT)));
```
Maybe you want to place line breaks to improve readability?
```suggestion
                    list.addToList(new Deadline(params[2], isDone,
                        LocalDateTime.parse(params[3], Duke.TIME_FORMAT)));
```

yezkez10 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,115 @@
+            }
+
+            storage.writeToFile(taskList);
+        }
```
I like the very clear logical flow here, but is there a way you can shorten the code here by abstracting each 'case' using the enum class?

yezkez10 commented on src/main/java/duke/exceptions/DukeDateTimeParseException.java:
```diff
@@ -0,0 +1,7 @@
+public class DukeDateTimeParseException extends DukeException {
+    public DukeDateTimeParseException() {
+        super("OOPS!!! Date and Time format does not match dd mmm yyyy - hh:mm");
+    }
```
Great abstraction of exception! Maybe you want to articulate the problem clearer by giving an example of a date and time that matches the format dd mm yyyy - hh:mm. Maybe the message can be:
"OOPS!!! Date and Time format invalid! Try with input format 30 Jul 2003 2:00"

yezkez10 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,102 @@
+        return s.toString();
+    }
+
+    @Override
```
Love that you have narrowed down and specified the equality for better precision!
Would be good to add JavaDoc to justify why the overriding equals methods are necessary, ie what make it different from the overarching equals method such that it warrants its own overriding method.

yezkez10 reviewed (commented):
LGTM👍 (looks GREAT to me)!
Overall, I like that the code was very OOP-ised. Excellent abstraction of methods into their relevant classes. On top of that, _phenomenal_ abstraction of exceptions and exception handling. 
If the circumstances allow it, you could try truncating the logic flow in getUserInput using the enum class, to make your code _even_ easier to read.
Great work thus far! 
"Be humble. Be hungry. And always be the hardest worker in the room." - Dwayne Johnson

AprupKale committed:
Move functionality to manage task list from Ui to Duke

AprupKale committed:
Update for A-Enums - Add enums for respective actions

AprupKale committed:
Fix checkstyle issues

AprupKale committed:
Fix issue with enum switch in Duke

AprupKale committed:
Set up JavaFX

AprupKale committed:
Update for Level 10 - Add GUI

AprupKale committed:
Merge branch 'branch-Level-10'

AprupKale committed:
Fix list display issue in GUI

AprupKale committed:
Fix checkstyle issues

AprupKale committed:
Update for A-Assertions - Add assertions

AprupKale committed:
Merge pull request #1 from AprupKale/branch-A-Assertions

Update for A-Assertions - Add assertions

AprupKale committed:
Merge branch 'master' into branch-A-CodeQuality

AprupKale committed:
Merge branch 'master' into branch-A-Streams

AprupKale committed:
Update for A-Streams - Use stream to filter tasks

AprupKale committed:
Merge pull request #2 from AprupKale/branch-A-Streams

Update for A-Streams - Use stream to filter tasks

AprupKale committed:
Merge branch 'master' into branch-A-CodeQuality

AprupKale committed:
Update for A-CodeQuality - Fix code to improve code quality

Fix code to improve naming and readability of code.

AprupKale committed:
Merge pull request #3 from AprupKale/branch-A-CodeQuality

Update for A-CodeQuality - Fix code to improve code quality

AprupKale committed:
Update for C-Priority - Add priority for tasks.

Add priority field for tasks.

Add functionality to change priority of tasks using
"change priority {index} {priority}" where {index}
is the index  of the task to modify in the list,
and {priority} is low, medium, or high.

AprupKale committed:
Merge branch 'branch-C-Priority'

AprupKale committed:
Update for A-CI - Set up Continuous Integration

Add .github/workflows/gradle.yml to set up continuous integration and
make GitHub Actions run the workflow specified every time certain
project events are triggered.

AprupKale committed:
Refactor Duke::getResponse to follow SLAP

Extracted multiple methods out of the single method getResponse
in Duke.

AprupKale committed:
Fix checkstyle issues with import ordering

AprupKale committed:
Update README.md

AprupKale committed:
Update for A-MoreTesting - Add Test classes for Ui and Duke

Add more JUnit tests to original test classes, and introduce
new JUnit tests with the new test classes.

AprupKale committed:
Update for A-MoreErrorHandling - Add DukeDateTimeOrderException

Throw exception when from time in an even is after the to time.

AprupKale committed:
Update for A-UserGuide - Add user guide and Ui demo image

AprupKale committed:
Fix file issues

Add functionality to create file if it does not exist.

AprupKale committed:
Fix cross-platform JavaFX issues
